tracing-core/src/collect.rs

```diff
@@ -436,14 +437,14 @@ pub trait Collect: 'static {
     /// # Safety
     ///
     /// The [`downcast_ref`] method expects that the pointer returned by
-    /// `downcast_raw` is non-null and points to a valid instance of the type
+    /// `downcast_raw` points to a valid instance of the type
     /// with the provided `TypeId`. Failure to ensure this will result in
     /// undefined behaviour, so implementing `downcast_raw` is unsafe.
     ///
     /// [`downcast_ref`]: #method.downcast_ref
-    unsafe fn downcast_raw(&self, id: TypeId) -> Option<*const ()> {
+    unsafe fn downcast_raw(&self, id: TypeId) -> Option<NonNull<()>> {
         if id == TypeId::of::<Self>() {
-            Some(self as *const Self as *const ())
+            Some(NonNull::new_unchecked(self as *const Self as *mut ()))
```

Comment:
nit, take it or leave it: `NonNull` [has a `From<&T>` impl](https://doc.rust-lang.org/stable/core/ptr/struct.NonNull.html#impl-From%3C%26%27_%20T%3E) which is essentially the same as `new_unchecked` except that it relies on an `&T` never being null as a language level invariant. So, we could alternatively express this as 
```suggestion
            Some(NonNull::from(self).cast::<()>())
```
which would get us the same behavior, but avoid the use of `new_unchecked`.

alternatively, if we are going to continue using `new_unchecked`, it would maybe be worth a safety comment explaining that we know the pointer will not be null, since it was constructed from an `&`-reference to `self`, which Rust guarantees is not null.